sumpy/fmm.py

```diff
@@ -483,6 +604,68 @@ def eval_direct(self, target_boxes, source_box_starts,
 
         return (pot, SumpyTimingFuture(self.queue, events))
 
+    @memoize_method
+    def multipole_to_local_precompute(self, src_rscale):
+        m2l_precomputed_exprs = self.m2l_precomputed_exprs_zeros()
+        events = []
+        for lev in range(self.tree.nlevels):
+            order = self.level_orders[lev]
+            precompute_kernel = \
+                self.code.m2l_optimized_precompute_kernel(order, order)
+
+            translation_classes_level_start, precomputed_exprs_view = \
+                    self.m2l_precomputed_exprs_view(m2l_precomputed_exprs, lev)
+
+            ntranslation_classes = precomputed_exprs_view.shape[0]
+
+            if ntranslation_classes == 0:
+                events.append([])
+                continue
+
+            m2l_translation_vectors = (
+                    self.translation_classes_data.m2l_translation_vectors())
+
+            evt, _ = precompute_kernel(
+                self.queue,
+                src_rscale=src_rscale,
+                translation_classes_level_start=translation_classes_level_start,
+                ntranslation_classes=ntranslation_classes,
+                m2l_precomputed_exprs=precomputed_exprs_view,
+                m2l_translation_vectors=m2l_translation_vectors,
+                ntranslation_vectors=m2l_translation_vectors.shape[1],
+                **self.kernel_extra_kwargs
+            )
+
+            events.append([evt])
+
+        return m2l_precomputed_exprs, events
+
+    def _use_multipole_to_local_precompute(self, kwargs_for_multipole_to_local,
```

Comment:
```suggestion
    def _add_m2l_precompute_kwargs(self, kwargs_for_m2l,
```

Reply:
Fixed in c2a8f519dcb29e18d79ad3fdffb9b640d7dfb7d7 and a67dbed770e63bf23bd3dac891f337b2805a2ac3